lib/nexthop_group.c

```diff
@@ -60,6 +60,16 @@ nexthop_group_cmd_compare(const struct nexthop_group_cmd *nhgc1,
 	return strcmp(nhgc1->name, nhgc2->name);
 }
 
+static struct nexthop *nexthop_group_tail(const struct nexthop_group *nhg)
+{
+	struct nexthop *nexthop = nhg->nexthop;
+
+	while (nexthop && nexthop->next)
+		nexthop = nexthop->next;
```

Comment:
I'm 80% sure what makes this entire ordeal slow isn't the comparisons, it's the fact that we're accessing the nexthop list, hopping from item to item in a sequential manner, taking a cache load latency each time.

This does nothing to address that.  Do you have perf data to confirm that it actually helps?

Reply:
I'm thinking that maybe the synthetic test produces sorted nexthops, so it is being helped because the 'next' nexthop almost always belongs at the end of the list?

Reply:
For the `add_sorted()` function I could switch to a binary search but it would still be O(n) since I have no easy way to find the midpoint and directly access it (I would have to iterate on the linked list at each level). @eqvinox Unless you have a cool trick you could show me?

The next best thing I think I can do is sort on the zapi nexthops array before the send.


@mjstapp I think you might be correct. Now that I think about it I bet sharpd does the same thing and that's what me and @donaldsharp were testing with.

So I think I will try to sort on the zapi nexthop array next preferably before the send.

Reply:
Would it make sense to do a sort on insertion here? It wouldn't be "more efficient" -- but it might spread the computation across more time?